fableous-fe/src/components/Canvas.tsx

```diff
@@ -18,417 +18,609 @@ import { ControllerRole, ToolMode, WSMessage, WSMessageType } from "../Data";
-            type: WSMessageType.Fill,
-            data: { x1: normX, y1: normY, color: targetColor },
-          } as WSMessage)
-        );
-      }
-    },
-    [role, wsRef]
-  );
+        ctx.putImageData(image, 0, 0);
+        if (role !== ControllerRole.Hub) {
+          const [normX, normY] = scaleDownXY(startX, startY);
+          wsRef.current?.send(
+            JSON.stringify({
+              role,
+              type: WSMessageType.Fill,
+              data: { x1: normX, y1: normY, color: targetColor },
+            } as WSMessage)
+          );
+        }
+      },
+      [canvasRef, role, scaleDownXY, wsRef]
+    );
 
-  const placeText = useCallback(
-    (x, y, message, fontSize) => {
-      if (!message) return;
+    const placeText = useCallback(
+      (x, y, id, text, fontSize) => {
+        const ctx = canvasRef.current.getContext(
+          "2d"
+        ) as CanvasRenderingContext2D;
+        ctx.font = `${fontSize * SCALE}px Arial`;
+        const [x1, y1, x2, y2] = getTextBounds(x, y, text, fontSize);
+        setTextShapes((prev) => ({
+          ...prev,
+          [id]: {
+            text,
+            fontSize,
+            x1,
+            y1,
+            x2,
+            y2,
+          } as TextShape,
+        }));
+        if (role !== ControllerRole.Hub) {
+          const [normX, normY] = scaleDownXY(x, y);
+          const [normFontSize] = scaleDownXY(fontSize, 0);
+          wsRef.current?.send(
+            JSON.stringify({
+              role,
+              type: WSMessageType.Text,
+              data: {
+                x1: normX,
+                y1: normY,
+                id,
+                text,
+                width: normFontSize,
+              },
+            } as WSMessage)
+          );
+        }
+      },
+      [canvasRef, getTextBounds, role, scaleDownXY, wsRef]
+    );
+
+    const refreshText = useCallback(() => {
       const ctx = canvasRef.current.getContext(
         "2d"
       ) as CanvasRenderingContext2D;
-      ctx.font = `${fontSize * SCALE}px Arial`;
-      ctx.textAlign = "center";
-      ctx.fillText(message, x, y);
-      if (role !== ControllerRole.Hub) {
-        const [normX, normY] = scaleDownXY(x, y);
-        const [normFontSize] = scaleDownXY(fontSize, 0);
-        wsRef.current?.send(
-          JSON.stringify({
-            role,
-            type: WSMessageType.Text,
-            data: {
-              x1: normX,
-              y1: normY,
-              width: normFontSize,
-              text: message,
-            },
-          } as WSMessage)
+      if (!ctx) return;
+      const { width, height } = canvasRef.current;
+      ctx.clearRect(0, 0, width, height);
+      Object.entries(textShapes).forEach(([id, shape]) => {
+        ctx.font = `${shape.fontSize * SCALE}px Arial`;
+        ctx.fillText(
+          shape.text,
+          (shape.x1 + shape.x2) / 2,
+          (shape.y1 + shape.y2) / 2
+        );
+        if (parseInt(id, 10) === editingText) {
```

Comment:
minor, but is editingTextId a more fitting name?

Reply:
done